daft/runners/ray_metrics.py

```diff
@@ -0,0 +1,218 @@
+                ray_task_id=ray_task_id,
+            )
+        )
+
+    def mark_task_end(self, execution_id: str, task_id: str, end: float):
+        # Add an EndTaskEvent
+        self._task_events[execution_id].append(EndTaskEvent(task_id=task_id, end=end))
+
+    def get_task_events(self, execution_id: str, idx: int) -> tuple[list[TaskEvent], int]:
+        events = self._task_events[execution_id]
+        return (events[idx:], len(events))
+
+    def collect_and_close(self, execution_id: str) -> dict[str, list[str]]:
+        """Collect the metrics associated with this execution, cleaning up the memory used for this execution ID"""
+        # Data about the available nodes and worker IDs in those nodes
+        node_data = self._node_info[execution_id].collect_node_info()
+
+        # Clean up the stats for this execution
+        del self._task_events[execution_id]
+        del self._node_info[execution_id]
+
+        return node_data
+
+
+@dataclasses.dataclass(frozen=True)
+class MetricsActorHandle:
+    execution_id: str
+    actor: ray.actor.ActorHandle
+
+    def wait(self) -> None:
+        """Call to block until the underlying actor is ready"""
+        return ray.get(self.actor.ready.remote())
+
+    def mark_task_start(
+        self,
+        task_id: str,
+        start: float,
+        node_id: str,
+        worker_id: str,
+        stage_id: int,
+        ray_assigned_resources: dict,
+        ray_task_id: str,
+    ) -> None:
+        self.actor.mark_task_start.remote(
+            self.execution_id,
+            task_id,
+            start,
+            node_id,
+            worker_id,
+            stage_id,
+            ray_assigned_resources,
+            ray_task_id,
+        )
+
+    def mark_task_end(
+        self,
+        task_id: str,
+        end: float,
+    ) -> None:
+        self.actor.mark_task_end.remote(
+            self.execution_id,
+            task_id,
+            end,
+        )
+
+    def get_task_events(self, idx: int) -> tuple[list[TaskEvent], int]:
+        """Collect task metrics from a given logical event index
+
+        Returns the task metrics and the new logical event index (to be used as a pagination offset token on subsequent requests)
+        """
+        return ray.get(self.actor.get_task_events.remote(self.execution_id, idx))
+
+    def collect_and_close(self) -> dict[str, set[str]]:
+        """Collect node metrics and close the metrics actor for this execution"""
+        return ray.get(self.actor.collect_and_close.remote(self.execution_id))
+
+
+# Creating/getting an actor from multiple threads is not safe.
+#
+# This could be a problem because our Scheduler does multithreaded executions of plans if multiple
+# plans are submitted at once.
+#
+# Pattern from Ray Data's _StatsActor:
+# https://github.com/ray-project/ray/blob/0b1d0d8f01599796e1109060821583e270048b6e/python/ray/data/_internal/stats.py#L447-L449
+_metrics_actor_lock: threading.RLock = threading.RLock()
```

Comment:
why you need reentrant lock? cant you use `threading.Lock `

Reply:
I believe this should be safe... Took this from Ray's code where they had RLock.

Reply:
Oh I meant you should't need a reentrant lock since theres no nested method calls. That will greatly speed up the locking speed.